File: Array/zhanchengguo/0702-有序链表转换二叉搜索树/Main.java
package com.test.ktapplication;


import android.util.Log;

public class Main {

    void main() {
        ListNode a = new ListNode(-10);
        ListNode b = new ListNode(-3);
        ListNode c = new ListNode(0);
        ListNode d = new ListNode(3);
        ListNode e = new ListNode(9);
        a.next = b;
        b.next = c;
        c.next = d;
        d.next = e;
        e.next = null;
        TreeNode result = sortedListToBST(a);
        Log.e("value is ", result + "");
    }

    public TreeNode sortedListToBST(ListNode head) {
        return buildTree(head, null);
    }


    public TreeNode buildTree(ListNode head, ListNode tail) {
        if (head == tail) {
            return null;
        }
        ListNode fast = head;
        ListNode slow = head;
        while (fast != tail && fast.next != tail) {
            fast = fast.next.next;
            slow = slow.next;
        }
        TreeNode root = new TreeNode(slow.val);
        root.left = buildTree(head, slow);
        root.right = buildTree(slow.next, tail);
        return root;
    }

}


class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

File: Array/zhanchengguo/0702-有序链表转换二叉搜索树/main.kt
package com.test.ktapplication

fun main() {
    val a = ListNode(-10)
    val b = ListNode(-3)
    val c = ListNode(0)
    val d = ListNode(3)
    val e = ListNode(9)
    a.next = b
    b.next = c
    c.next = d
    d.next = e
    e.next = null
    val result = sortedListToBST(a)
    print("value is ${result}\n")
}

fun sortedListToBST(head: ListNode): TreeNode? {
    return buildTree(head, null)
}


fun buildTree(head: ListNode?, tail: ListNode?): TreeNode? {
    if (head === tail) {
        return null
    }
    var fast = head
    var slow = head
    while (fast !== tail && fast?.next !== tail) {
        fast = fast?.next?.next
        slow = slow?.next
    }
    val root = TreeNode(slow?.`val`)
    root.left = buildTree(head, slow)
    root.right = buildTree(slow?.next, tail)
    return root
}


class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

class TreeNode(var `val`: Int?) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}